src/minimap.cpp

```diff
@@ -120,89 +120,76 @@ void MinimapUpdateThread::doUpdate()
-
-	for (s16 i = blockpos_max.Y; i > blockpos_min.Y - 1; i--) {
-		std::map<v3s16, MinimapMapblock *>::iterator it =
-			m_blocks_cache.find(v3s16(blockpos_max.X, i, blockpos_max.Z));
-		if (it != m_blocks_cache.end()) {
-			MinimapMapblock *mmblock = it->second;
-			MinimapPixel *pixel = &mmblock->data[relpos.Z * MAP_BLOCKSIZE + relpos.X];
-			air_count += pixel->air_count;
-		}
-	}
-
-	return air_count;
+	return v3s16(std::max(a.X, b.X), std::max(a.Y, b.Y), std::max(a.Z, b.Z));
 }
 
-void MinimapUpdateThread::getMap(v3s16 pos, s16 size, s16 height, bool is_radar)
+void MinimapUpdateThread::getMap(v3s16 pos, s16 size, s16 height)
 {
-	v3s16 p = v3s16(pos.X - size / 2, pos.Y, pos.Z - size / 2);
-
-	for (s16 x = 0; x < size; x++)
-	for (s16 z = 0; z < size; z++) {
-		MapNode n(CONTENT_AIR);
-		MinimapPixel *mmpixel = &data->minimap_scan[x + z * size];
-
-		if (!is_radar) {
-			s16 pixel_height = 0;
-			MinimapPixel *cached_pixel =
-				getMinimapPixel(v3s16(p.X + x, p.Y, p.Z + z), height, &pixel_height);
-			if (cached_pixel) {
-				n = cached_pixel->n;
-				mmpixel->height = pixel_height;
+	v3s16 region(size, 0, size);
+	v3s16 pos_min(pos.X - size / 2, pos.Y - height / 2, pos.Z - size / 2);
+	v3s16 pos_max(pos_min.X + size - 1, pos.Y + height / 2, pos_min.Z + size - 1);
+	v3s16 blockpos_min = getNodeBlockPos(pos_min);
+	v3s16 blockpos_max = getNodeBlockPos(pos_max);
+
+// clear the map
+	for (int z = 0; z < size; z++)
+	for (int x = 0; x < size; x++)
+	{
+		MinimapPixel &mmpixel = data->minimap_scan[x + z * size];
+		mmpixel.air_count = 0;
+		mmpixel.height = 0;
+		mmpixel.n = MapNode(CONTENT_AIR);
+	}
+
+// draw the map
+	v3s16 blockpos;
+	for (blockpos.Z = blockpos_min.Z; blockpos.Z <= blockpos_max.Z; ++blockpos.Z)
+	for (blockpos.Y = blockpos_min.Y; blockpos.Y <= blockpos_max.Y; ++blockpos.Y)
+	for (blockpos.X = blockpos_min.X; blockpos.X <= blockpos_max.X; ++blockpos.X)
+	{
+		std::map<v3s16, MinimapMapblock *>::const_iterator pblock =
+			m_blocks_cache.find(blockpos);
+		if (pblock == m_blocks_cache.end())
+			continue;
+		const MinimapMapblock &block = *pblock->second;
+
+		v3s16 block_node_min(blockpos * MAP_BLOCKSIZE);
+		v3s16 block_node_max(block_node_min + MAP_BLOCKSIZE - 1);
+		// clip
+		v3s16 range_min = componentwise_max(block_node_min, pos_min);
+		v3s16 range_max = componentwise_min(block_node_max, pos_max);
+
+		v3s16 pos;
+		pos.Y = range_min.Y;
+		for (pos.Z = range_min.Z; pos.Z <= range_max.Z; ++pos.Z)
+		for (pos.X = range_min.X; pos.X <= range_max.X; ++pos.X)
+		{
+			v3s16 inblock_pos = pos - block_node_min;
+			const MinimapPixel &in_pixel =
+				block.data[inblock_pos.Z * MAP_BLOCKSIZE + inblock_pos.X];
+
+			v3s16 inmap_pos = pos - pos_min;
+			MinimapPixel &out_pixel =
+				data->minimap_scan[inmap_pos.X + inmap_pos.Z * size];
+
+			out_pixel.air_count += in_pixel.air_count;
+			if (in_pixel.n.param0 != CONTENT_AIR) {
+				out_pixel.n = in_pixel.n;
+				out_pixel.height = inmap_pos.Y + in_pixel.height;
```

Comment:
`inmap_pos.Y + in_pixel.height` is the pixel Y, based on the lowest possible position.
Old code was doing `height -= MAP_BLOCKSIZE` so that `height + in_pixel.height` was the same (maybe plus a constant, that does not matter).